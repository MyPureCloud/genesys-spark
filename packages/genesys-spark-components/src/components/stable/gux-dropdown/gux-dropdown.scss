@mixin gux-input-focus-border {
  outline: var(--gse-ui-formControl-input-focus-border-width)
    var(--gse-ui-formControl-input-focus-border-style)
    var(--gse-ui-formControl-input-focus-border-color);
  outline-offset: var(--gse-semantic-focusRing-offset);
}

:host {
  box-sizing: border-box;
  color: var(--gse-ui-formControl-input-populatedColor);
}

.gux-field,
.gux-target-container-expanded {
  all: unset;
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  place-content: stretch center;
  align-items: center;
  width: 100%;
  height: var(--gse-ui-formControl-input-textfield-height);
  font-family: var(--gse-ui-formControl-input-contentText-fontFamily);
  font-size: var(--gse-ui-formControl-input-contentText-fontSize);
  font-weight: var(--gse-ui-formControl-input-contentText-fontWeight);
  line-height: var(--gse-ui-formControl-input-contentText-lineHeight);
  cursor: pointer;
  background-color: var(--gse-ui-formControl-input-backgroundColor);
}

.gux-target-container-expanded,
.gux-target-container-collapsed .gux-field {
  padding: var(--gse-ui-formControl-input-padding);
}

.gux-error {
  &.gux-target-container-collapsed .gux-field-button,
  &.gux-target-container-expanded {
    border: var(--gse-ui-formControl-input-error-border-width)
      var(--gse-ui-formControl-input-error-border-style)
      var(--gse-ui-formControl-input-error-border-color);
  }
}

.gux-disabled {
  &.gux-target-container-collapsed .gux-field-button,
  &.gux-target-container-expanded {
    border: var(--gse-ui-formControl-input-disabled-border-width)
      var(--gse-ui-formControl-input-disabled-border-style)
      var(--gse-ui-formControl-input-disabled-border-color);
  }
}

.gux-target-container-collapsed .gux-field-button:hover,
.gux-target-container-expanded:hover {
  border: var(--gse-ui-formControl-input-hover-border-width)
    var(--gse-ui-formControl-input-hover-border-style)
    var(--gse-ui-formControl-input-hover-border-color);
}

.gux-field {
  &.gux-input-field {
    height: var(--gse-ui-formControl-input-contentText-lineHeight);
  }

  .gux-field-content {
    // COMUI-1650 consider updating default z-index values
    // Default value for popup is 2, and this tooltip must show on top
    --gux-zindex-tooltip: 3;

    display: flex;
    flex: 1 1 0;
    flex-direction: row;
    flex-wrap: nowrap;
    place-content: stretch flex-start;
    align-items: flex-start;
    min-width: 0;
    height: var(--gse-ui-formControl-input-contentText-lineHeight);

    .gux-filter,
    .gux-selected-option,
    .gux-placeholder {
      flex: 1 1 auto;
      align-self: auto;
      order: 0;
      padding: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .gux-filter {
      position: relative;
      height: 100%;

      .gux-filter-input {
        all: unset;
        width: 100%;
        color: transparent;
        caret-color: var(--gse-ui-formControl-input-populatedColor);
      }

      .gux-filter-display {
        white-space: pre;

        .gux-filter-text {
          color: var(--gse-ui-formControl-input-populatedColor);
        }

        .gux-filter-suggestion {
          color: var(--gse-ui-formControl-input-suggestionColor);
        }
      }

      .gux-filter-input,
      .gux-filter-display {
        position: absolute;
      }
    }

    .gux-placeholder {
      color: var(--gse-ui-formControl-input-placeholderColor);
    }
  }

  .gux-expand-icon {
    flex: 0 0 auto;
    align-self: auto;
    order: 0;
    width: var(--gse-ui-formControl-input-inputIcon-size);
    height: var(--gse-ui-formControl-input-inputIcon-size);
    padding-left: var(--gse-ui-dropdown-gap);
    color: var(--gse-ui-formControl-input-inputIcon-iconEndColor);
  }
}

.gux-target-container-expanded {
  border: var(--gse-ui-formControl-input-active-border-width)
    var(--gse-ui-formControl-input-active-border-style)
    var(--gse-ui-formControl-input-active-border-color);
  border-radius: var(--gse-ui-formControl-input-borderRadius);

  &:focus-visible {
    @include gux-input-focus-border;
  }

  &:focus-within:has(:focus-visible) {
    @include gux-input-focus-border;
  }

  .gux-filter-input {
    background-color: inherit;
    border: none;

    &:focus {
      border: none;
      outline: none;
    }
  }

  .gux-field-button {
    width: auto;
    height: var(--gse-ui-formControl-input-contentText-lineHeight);
    margin: 0;
    background: inherit;
    border: none;
    outline: none;
    box-shadow: none;

    &:focus {
      outline: none;
    }
  }
}

.gux-target-container-collapsed .gux-field-button {
  border: var(--gse-ui-formControl-input-default-border-width)
    var(--gse-ui-formControl-input-default-border-style)
    var(--gse-ui-formControl-input-default-border-color);
  border-radius: var(--gse-ui-formControl-input-borderRadius);

  &:focus-visible {
    @include gux-input-focus-border;
  }

  &:focus-within:has(:focus-visible) {
    @include gux-input-focus-border;
  }
}

::slotted(gux-listbox) {
  outline: none;
  box-shadow: var(--gse-ui-menu-boxShadow-x) var(--gse-ui-menu-boxShadow-y)
    var(--gse-ui-menu-boxShadow-blur) var(--gse-ui-menu-boxShadow-spread)
    var(--gse-ui-menu-boxShadow-color);
}

// Selected option styles

.gux-selected-icon {
  display: flex;
  align-items: center;

  gux-icon {
    width: var(--gse-ui-formControl-input-inputIcon-size);
    height: var(--gse-ui-formControl-input-inputIcon-size);
    padding-right: var(--gse-ui-dropdown-gap);
  }
}
