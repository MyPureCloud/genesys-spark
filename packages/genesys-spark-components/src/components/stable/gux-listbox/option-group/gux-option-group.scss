@use '~genesys-spark/dist/beta/tokens/flare/ui.scss';
@use '~genesys-spark/dist/scss/typography.scss';

.gux-option-group {
  display: block;
  flex: 1 1 auto;
  align-self: auto;

  .gux-option-group-label {
    display: block;
    padding: ui.$gse-ui-menu-groupedMenu-title-padding;
    color: ui.$gse-ui-menu-groupedMenu-title-foregroundColor;

    @include typography.heading-overline;
  }
}

:host(.gux-filtered) {
  display: none;
}
