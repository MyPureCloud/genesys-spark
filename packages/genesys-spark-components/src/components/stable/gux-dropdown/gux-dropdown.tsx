import {
  Component,
  Element,
  Event,
  EventEmitter,
  forceUpdate,
  h,
  Listen,
  JSX,
  Prop,
  State,
  Watch
} from '@stencil/core';

import { OnClickOutside } from '@utils/decorator/on-click-outside';
import { buildI18nForComponent, GetI18nValue } from '../../../i18n';
import simulateNativeEvent from '@utils/dom/simulate-native-event';
import { calculateInputDisabledState } from '@utils/dom/calculate-input-disabled-state';
import { onInputDisabledStateChange } from '@utils/dom/on-input-disabled-state-change';
import { afterNextRender } from '@utils/dom/after-next-render';
import { trackComponent } from '@utils/tracking/usage';
import { OnMutation } from '@utils/decorator/on-mutation';

import translationResources from './i18n/en.json';

import {
  getSearchOption,
  getListOptions,
  setInitialActiveOption
} from '../gux-listbox/gux-listbox.service';
import {
  ListboxOptionElement,
  ValidOptionTag
} from '../gux-listbox/options/option-types';
import { GuxFilterTypes } from './gux-dropdown.types';
import { hasSlot } from '@utils/dom/has-slot';

/**
 * Our Dropdown component. In the most basic case, it's used with `gux-option` to give users
 * a list of text options to select from, but other types of options with different appearance
 * can be created by creating a new component and adding it to `validOptionTags` list in
 * gux-dropdown-types.ts, then following the resulting compiler errors.
 *
 * @slot - for a gux-listbox containing ValidDropdownOption children
 */
@Component({
  styleUrl: 'gux-dropdown.scss',
  tag: 'gux-dropdown',
  shadow: { delegatesFocus: true }
})
export class GuxDropdown {
  private i18n: GetI18nValue;
  private fieldButtonElement: HTMLElement;
  private filterElement: HTMLInputElement;
  private listboxElement: HTMLGuxListboxElement;
  private truncateElement: HTMLGuxTruncateElement;

  @Element()
  private root: HTMLGuxDropdownElement;

  @Prop({ mutable: true })
  value: string;

  @Prop()
  disabled: boolean = false;

  @Prop()
  required: boolean = false;

  @Prop()
  loading: boolean = false;

  @Prop()
  placeholder: string;

  @Prop()
  filterType: GuxFilterTypes = 'none';

  @Prop()
  hasError: boolean = false;

  /**
   * allows dropdown popup to be wider than input
   * defaults to fitting content if width is not specified for listbox
   * default min-width is set to width of input
   */
  @Prop()
  exceedTargetWidth: boolean = false;

  @State()
  private expanded: boolean = false;

  @State()
  private filter: string = '';

  @Watch('expanded')
  watchExpanded(expanded: boolean) {
    if (!expanded) {
      this.filter = '';
    }
  }

  @Watch('value')
  watchValue(newValue: string) {
    this.validateValue(newValue, this.listboxElement);
  }

  @Watch('filter')
  handleFilter(filter: string) {
    this.guxfilter.emit(filter);
  }

  @Watch('disabled')
  watchDisabled(disabled: boolean) {
    if (disabled) {
      this.expanded = false;
    }
  }

  @Listen('keydown')
  onKeydown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Escape':
        if (this.isFilterable()) {
          if (document.activeElement === this.listboxElement) {
            return this.filterElement.focus();
          }
        }
        this.collapseListbox('focusFieldButton');
        return;
      case 'Tab':
        if (this.shiftTabFromFilterListbox(event)) {
          event.preventDefault();
          return this.filterElement.focus();
        } else if (this.shiftTabFromExpandedFilterInput(event)) {
          event.preventDefault();
          return this.collapseListbox('focusFieldButton');
        } else {
          this.collapseListbox('noFocusChange');
        }
        return;
      case 'ArrowDown':
        if (this.activeElementNotListbox()) {
          event.preventDefault();
          this.expanded = true;
          setInitialActiveOption(this.listboxElement);
        }
        return;
    }
  }

  @Listen('internallistboxoptionsupdated')
  onInternallistboxoptionsupdated(event: CustomEvent): void {
    event.stopPropagation();
    forceUpdate(this.root);
  }

  @Listen('blur')
  onBlur(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @Listen('focus')
  onFocus(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @Listen('focusout')
  onFocusout(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @Listen('focusin')
  onFocusin(event: FocusEvent): void {
    this.stopPropagationOfInternalFocusEvents(event);
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  onClickOutside() {
    this.collapseListbox('noFocusChange');
  }

  @Event()
  guxexpanded: EventEmitter<void>;

  @Event()
  guxcollapsed: EventEmitter<void>;

  @Event()
  private guxfilter: EventEmitter<string>;

  @OnMutation({ childList: true, subtree: true })
  onMutation(): void {
    if (this.listboxElement) {
      return;
    }

    this.listboxElement = this.root?.querySelector('gux-listbox');
    this.applyListboxEventListeners();
  }

  @Listen('internalexpanded')
  onInternalExpanded(event: CustomEvent): void {
    event.stopPropagation();
    this.guxexpanded.emit();

    if (this.listboxElement) {
      afterNextRender(() => {
        this.listboxElement.focus();

        if (this.isFilterable() && this.filterElement) {
          this.filterElement.focus();
        }
      });
    }
  }

  @Listen('internalcollapsed')
  onInternalCollapsed(event: CustomEvent): void {
    event.stopPropagation();
    this.guxcollapsed.emit();
  }

  connectedCallback(): void {
    this.listboxElement = this.root?.querySelector('gux-listbox');
    if (this.listboxElement) {
      this.validateValue(this.value, this.listboxElement);
    }
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
    this.i18n = await buildI18nForComponent(this.root, translationResources);

    onInputDisabledStateChange(this.root, () => {
      forceUpdate(this.root);
    });
  }

  componentDidLoad(): void {
    this.applyListboxEventListeners();
  }

  componentWillRender(): void {
    if (this.listboxElement) {
      this.validateValue(this.value, this.listboxElement);
      this.listboxElement.loading = this.loading;
      this.listboxElement.filterType = this.filterType;
      this.listboxElement.filter = this.filter;
    }
  }
  private showTooltip(): void {
    void this.truncateElement?.setShowTooltip();
  }

  private hideTooltip(): void {
    void this.truncateElement?.setHideTooltip();
  }

  private validateValue(
    newValue: string,
    listboxElement: HTMLGuxListboxElement
  ): void {
    if (newValue === undefined) {
      if (listboxElement) {
        listboxElement.value = newValue;
      }
      return;
    }

    const selectedListboxOptionElement = this.getOptionElementByValue(newValue);

    if (selectedListboxOptionElement) {
      listboxElement.value = newValue;
      return;
    }
  }

  private applyListboxEventListeners(): void {
    this.listboxElement?.addEventListener('input', (event: InputEvent) => {
      event.stopPropagation();

      this.updateValue((event.target as HTMLGuxListboxElement).value);
    });
    this.listboxElement?.addEventListener('change', (event: InputEvent) => {
      event.stopPropagation();
    });
  }

  private stopPropagationOfInternalFocusEvents(event: FocusEvent): void {
    if (this.root.contains(event.relatedTarget as Node)) {
      return event.stopImmediatePropagation();
    }
  }

  private isFilterable() {
    return this.filterType === 'starts-with' || this.filterType === 'custom';
  }

  get optionElements(): Array<ListboxOptionElement> {
    return getListOptions(this.listboxElement);
  }

  private getOptionElementByValue(value: string): HTMLElement {
    return this.optionElements.find(optionElement => {
      return optionElement.value === value;
    });
  }

  private fieldButtonClick(): void {
    this.expanded = !this.expanded;
  }

  private filterInput(event: InputEvent): void {
    event.stopPropagation();
    this.filter = this.filterElement.value;
  }

  private shiftTabFromExpandedFilterInput(event: KeyboardEvent): boolean {
    return (
      event.shiftKey &&
      this.isFilterable() &&
      this.expanded &&
      !(document.activeElement === this.listboxElement)
    );
  }

  private shiftTabFromFilterListbox(event: KeyboardEvent): boolean {
    return (
      event.shiftKey &&
      this.isFilterable() &&
      document.activeElement === this.listboxElement
    );
  }

  private activeElementNotListbox(): boolean {
    return document.activeElement !== this.listboxElement;
  }

  private filterKeydown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'ArrowDown':
        event.stopImmediatePropagation();
        event.preventDefault();
        this.listboxElement.focus();
        setInitialActiveOption(this.listboxElement);
        return;
      case 'Enter':
        void this.listboxElement.guxSelectActive();
        event.preventDefault();
        return;
    }
  }

  private filterKeyup(event: KeyboardEvent): void {
    switch (event.key) {
      case ' ':
        event.preventDefault();
        return;
    }
  }

  private collapseListbox(
    focusChange: 'noFocusChange' | 'focusFieldButton'
  ): void {
    if (this.expanded) {
      this.expanded = false;
    }

    if (focusChange === 'focusFieldButton') {
      this.fieldButtonElement.focus();
    }
  }

  private updateValue(newValue: string): void {
    if (this.value !== newValue) {
      this.value = newValue;
      simulateNativeEvent(this.root, 'input');
      simulateNativeEvent(this.root, 'change');
    }
    this.collapseListbox('focusFieldButton');
  }

  private getTypeaheadText(filter: string): string {
    const filterLength = filter.length;
    if (filterLength > 0 && !this.loading) {
      const option = getSearchOption(this.listboxElement, filter);
      if (option && this.filterType !== 'custom') {
        //The text content needs to be trimmed as white space can occur around the textContent if options are populated asynchronously.
        return option.textContent.trim().substring(filterLength);
      }
    }

    return '';
  }

  private renderTargetDisplay(): JSX.Element {
    const selectedListboxOptionElement = this.getOptionElementByValue(
      this.value
    );
    if (selectedListboxOptionElement) {
      return (
        <div class="gux-selected-option">
          {this.renderSelectedItem(selectedListboxOptionElement)}
        </div>
      ) as JSX.Element;
    }

    return (
      <div class="gux-placeholder">
        {this.placeholder || this.i18n('noSelection')}
      </div>
    ) as JSX.Element;
  }

  /**
   * Renders the selection display for the selected item. This function needs a branch to handle
   * each type defined in GuxDropdownOptionType
   *
   * @param item The selected item. This can be any of the node types defined in GuxDropdownOptionType.
   * @returns Rendered selection details.
   */
  private renderSelectedItem(item: HTMLElement): JSX.Element {
    const tag = item.tagName.toLowerCase() as ValidOptionTag;
    switch (tag) {
      case 'gux-option':
        return this.renderOption(item as HTMLGuxOptionElement);
      case 'gux-option-icon':
        return this.renderIconOption(item as HTMLGuxOptionIconElement);
      default:
        // eslint-disable-next-line no-case-declarations
        const _exhaustiveCheck: never = tag;
        return _exhaustiveCheck;
    }
  }

  private renderOption(option: HTMLGuxOptionElement): JSX.Element {
    let optionText = option.textContent;
    if (hasSlot(option, 'subtext')) {
      const subtext = option.querySelector('[slot=subtext]');
      optionText = optionText.substring(
        0,
        optionText.length - subtext.textContent.length
      );
    }
    return (
      <gux-truncate ref={el => (this.truncateElement = el)} dir="auto">
        {optionText}
      </gux-truncate>
    ) as JSX.Element;
  }

  private renderIconOption(iconOption: HTMLGuxOptionIconElement): JSX.Element {
    let iconStyle = null;
    if (iconOption.iconColor !== null) {
      iconStyle = { color: iconOption.iconColor };
    }
    return (
      <span class="gux-selected-icon">
        <gux-icon
          icon-name={iconOption.iconName}
          style={iconStyle}
          decorative
        ></gux-icon>
        <gux-truncate ref={el => (this.truncateElement = el)}>
          {iconOption.textContent}
        </gux-truncate>
      </span>
    ) as JSX.Element;
  }

  private renderFilterInputField(): JSX.Element {
    if (this.expanded && this.isFilterable()) {
      return (
        <div class="gux-field gux-input-field" dir="auto">
          <div class="gux-field-content">
            <div class="gux-filter">
              <div class="gux-filter-display">
                <span class="gux-filter-text">{this.filter}</span>
                <span class="gux-filter-suggestion">
                  {this.getTypeaheadText(this.filter)}
                </span>
              </div>
              <div class="input-and-dropdown-button">
                <input
                  onClick={this.fieldButtonClick.bind(this)}
                  class="gux-filter-input"
                  type="text"
                  aria-label={this.i18n('filterResults')}
                  ref={el => (this.filterElement = el)}
                  onInput={this.filterInput.bind(this)}
                  onKeyDown={this.filterKeydown.bind(this)}
                  onKeyUp={this.filterKeyup.bind(this)}
                  disabled={this.disabled}
                ></input>
              </div>
            </div>
          </div>
        </div>
      ) as JSX.Element;
    }
  }

  private renderPopup(): JSX.Element {
    return (<slot slot="popup"></slot>) as JSX.Element;
  }

  private renderTarget(): JSX.Element {
    return (
      <div
        class={{
          'gux-target-container-expanded': this.expanded && this.isFilterable(),
          'gux-target-container-collapsed': !(
            this.expanded && this.isFilterable()
          ),
          'gux-error': this.hasError,
          'gux-disabled': this.disabled
        }}
        slot="target"
      >
        {this.renderFilterInputField()}
        <button
          type="button"
          class="gux-field gux-field-button"
          disabled={calculateInputDisabledState(this.root)}
          onClick={this.fieldButtonClick.bind(this)}
          onFocusin={this.showTooltip.bind(this)}
          onFocusout={this.hideTooltip.bind(this)}
          ref={el => (this.fieldButtonElement = el)}
          aria-haspopup="listbox"
          aria-expanded={this.expanded.toString()}
        >
          {this.renderTargetContent()}
          {this.renderRadialLoading()}
          <gux-icon
            class={{
              'gux-expand-icon': true
            }}
            screenreader-text={this.i18n('dropdown')}
            iconName={
              this.expanded
                ? 'custom/chevron-up-small-regular'
                : 'custom/chevron-down-small-regular'
            }
          ></gux-icon>
        </button>
      </div>
    ) as JSX.Element;
  }
  private renderTargetContent(): JSX.Element {
    if (!(this.expanded && this.isFilterable())) {
      return (
        <div class="gux-field-content">{this.renderTargetDisplay()}</div>
      ) as JSX.Element;
    }
  }

  private renderRadialLoading(): JSX.Element {
    if (this.loading && !this.expanded) {
      return (
        <gux-radial-loading context="input"></gux-radial-loading>
      ) as JSX.Element;
    }
  }

  render(): JSX.Element {
    return (
      <gux-popup
        expanded={this.expanded && (!this.loading || this.isFilterable())}
        disabled={this.disabled || (this.loading && !this.isFilterable())}
        exceedTargetWidth={this.exceedTargetWidth}
      >
        {this.renderTarget()}
        {this.renderPopup()}
      </gux-popup>
    ) as JSX.Element;
  }
}
