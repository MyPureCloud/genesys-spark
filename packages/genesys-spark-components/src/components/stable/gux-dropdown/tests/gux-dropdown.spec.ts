import { newSpecPage } from '@test/specTestUtils';

import { GuxDropdown } from '../gux-dropdown';
import { GuxListbox } from '../../gux-listbox/gux-listbox';
import { GuxOption } from '../../gux-listbox/options/gux-option/gux-option';
import { GuxOptionIcon } from '../../gux-listbox/options/gux-option-icon/gux-option-icon';
import { GuxOptionGroup } from '../../gux-listbox/option-group/gux-option-group';

const language = 'en';

describe('gux-dropdown', () => {
  describe('#render', () => {
    it(`should render as expected with gux-option`, async () => {
      const components = [GuxDropdown, GuxListbox, GuxOption];
      const html = `
      <gux-dropdown lang="en" value="j">
        <gux-listbox aria-label="Animals">
          <gux-option value="a">Ant<span slot="subtext">Small</span></gux-option>
          <gux-option value="b">Bat</gux-option>
          <gux-option value="c">Cat</gux-option>
          <gux-option value="d">Dog</gux-option>
          <gux-option value="e">Eel</gux-option>
          <gux-option value="f">Frog</gux-option>
          <gux-option value="g">Goat</gux-option>
          <gux-option value="h">Horse<span slot="subtext">Large</span></gux-option>
          <gux-option value="i">Ibis</gux-option>
        </gux-listbox>
      </gux-dropdown>
      `;
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdown);
      expect(page.root).toMatchSnapshot();
    });

    it(`should render as expected with gux-option-icon`, async () => {
      const components = [GuxDropdown, GuxListbox, GuxOptionIcon];
      const html = `
      <gux-dropdown lang="en" value="j">
        <gux-listbox aria-label="Animals">
          <gux-option-icon
          icon-name="user"
          icon-color="var(--gux-blue-60)"
          value="leonardo"
          >Leonardo<span slot="subtext">TMNT</span></gux-option-icon
        >
        <gux-option-icon icon-name="user" icon-color="#8452cf" value="donatello"
          >Donatello</gux-option-icon
        >
        <gux-option-icon
          icon-name="user"
          icon-color="rgb(234, 11, 11)"
          value="raphael"
          >Raphael</gux-option-icon
        >
        <gux-option-icon
          icon-name="user"
          icon-color="#e08915"
          icon-sr-text="Screenreader information about the icon goes here, if needed"
          value="michelangelo"
          >Michelangelo</gux-option-icon
        >
        </gux-listbox>
      </gux-dropdown>
      `;
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdown);
      expect(page.root).toMatchSnapshot();
    });

    it(`should render as expected with gux-option-group and gux-option`, async () => {
      const components = [GuxDropdown, GuxListbox, GuxOption, GuxOptionGroup];
      const html = `
      <gux-dropdown lang="en" value="f">
        <gux-listbox aria-label="Animals">
          <gux-option-group-beta label="Group 1">
            <gux-option value="a">Ant</gux-option>
            <gux-option value="b">Bat</gux-option>
            <gux-option value="c" disabled>Cat</gux-option>
            <gux-option value="d">Dog</gux-option>
          </gux-option-group-beta>
          <gux-option-group-beta label="Group 2">
            <gux-option value="e">Eel</gux-option>
            <gux-option value="f">Frog</gux-option>
            <gux-option value="g">Goat</gux-option>
          </gux-option-group-beta>
          <gux-option-group-beta label="Group 3">
            <gux-option value="h">Horse</gux-option>
            <gux-option value="i">Ibis</gux-option>
          </gux-option-group-beta>
        </gux-listbox>
      </gux-dropdown>
      `;
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdown);
      expect(page.root).toMatchSnapshot();
    });

    it(`should render as expected with gux-option-group and gux-option-icon`, async () => {
      const components = [
        GuxDropdown,
        GuxListbox,
        GuxOptionIcon,
        GuxOptionGroup
      ];
      const html = `
      <gux-dropdown lang="en" value="f">
        <gux-listbox aria-label="Animals">
          <gux-option-group-beta label="Group 1">
            <gux-option-icon
            icon-name="user"
            icon-color="var(--gux-blue-60)"
            value="leonardo"
            >Leonardo<span slot="subtext">TMNT</span></gux-option-icon
          >
          <gux-option-icon icon-name="user" icon-color="#8452cf" value="donatello"
            >Donatello</gux-option-icon
          >
          </gux-option-group-beta>
          <gux-option-group-beta label="Group 2">
            <gux-option-icon
              icon-name="user"
              icon-color="rgb(234, 11, 11)"
              value="raphael"
              >Raphael</gux-option-icon
            >
            <gux-option-icon
              icon-name="user"
              icon-color="#e08915"
              icon-sr-text="Screenreader information about the icon goes here, if needed"
              value="michelangelo"
              >Michelangelo</gux-option-icon
            >
          </gux-option-group-beta>
        </gux-listbox>
      </gux-dropdown>
      `;
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdown);
      expect(page.root).toMatchSnapshot();
    });
    it(`should render as expected with gux-option whose slotted content includes whitespace`, async () => {
      const components = [GuxDropdown, GuxListbox, GuxOption];
      // With frog option selected below, the snapshot should show the selected item's text with
      // dropdown closed as "Frog"
      // Prior bug only considered first node of slot content, which below is a whitespace text node.
      // And not "Frog Amphibian" because subtext slot should be ignored.
      const html = `
      <gux-dropdown lang="en" value="f">
        <gux-listbox aria-label="Animals">
          <gux-option value="e">Eel</gux-option>
          <gux-option value="f">
            <span>Frog</span>
            <span slot="subtext">Amphibian</span>
          </gux-option>
          <gux-option value="g">Goat</gux-option>
        </gux-listbox>
      </gux-dropdown>
      `;
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdown);
      expect(page.root).toMatchSnapshot();
    });
  });
});
