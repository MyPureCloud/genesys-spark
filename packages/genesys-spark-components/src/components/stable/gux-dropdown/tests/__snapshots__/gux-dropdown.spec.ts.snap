// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-dropdown #render should render as expected with gux-option 1`] = `
<gux-dropdown lang="en" value="j">
  <mock:shadow-root>
    <gux-popup>
      <div class="gux-target-container-collapsed" slot="target">
        <button aria-expanded="false" aria-haspopup="listbox" class="gux-field gux-field-button" type="button">
          <div class="gux-field-content">
            <div class="gux-placeholder">
              Select...
            </div>
          </div>
          <gux-icon class="gux-expand-icon" iconname="chevron-small-down" screenreader-text="Dropdown"></gux-icon>
        </button>
      </div>
      <slot slot="popup"></slot>
    </gux-popup>
  </mock:shadow-root>
  <gux-listbox aria-label="Animals" role="listbox" tabindex="0">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="a">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Ant
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="b">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Bat
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="c">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Cat
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="d">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Dog
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="e">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Eel
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="f">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Frog
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="g">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Goat
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="h">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Horse
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="i">
      <mock:shadow-root>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Ibis
    </gux-option>
  </gux-listbox>
</gux-dropdown>
`;

exports[`gux-dropdown #render should render as expected with gux-option-icon 1`] = `
<gux-dropdown lang="en" value="j">
  <mock:shadow-root>
    <gux-popup>
      <div class="gux-target-container-collapsed" slot="target">
        <button aria-expanded="false" aria-haspopup="listbox" class="gux-field gux-field-button" type="button">
          <div class="gux-field-content">
            <div class="gux-placeholder">
              Select...
            </div>
          </div>
          <gux-icon class="gux-expand-icon" iconname="chevron-small-down" screenreader-text="Dropdown"></gux-icon>
        </button>
      </div>
      <slot slot="popup"></slot>
    </gux-popup>
  </mock:shadow-root>
  <gux-listbox aria-label="Animals" role="listbox" tabindex="0">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="var(--gux-blue-60)" icon-name="user" id="gux-option-icon-i" role="option" value="leonardo">
      <mock:shadow-root>
        <gux-icon decorative="" icon-name="user" style="color: var(--gux-blue-60);"></gux-icon>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Leonardo
    </gux-option-icon>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="#8452cf" icon-name="user" id="gux-option-icon-i" role="option" value="donatello">
      <mock:shadow-root>
        <gux-icon decorative="" icon-name="user" style="color: #8452cf;"></gux-icon>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Donatello
    </gux-option-icon>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="rgb(234, 11, 11)" icon-name="user" id="gux-option-icon-i" role="option" value="raphael">
      <mock:shadow-root>
        <gux-icon decorative="" icon-name="user" style="color: rgb(234, 11, 11);"></gux-icon>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Raphael
    </gux-option-icon>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="#e08915" icon-name="user" icon-sr-text="Screenreader information about the icon goes here, if needed" id="gux-option-icon-i" role="option" value="michelangelo">
      <mock:shadow-root>
        <gux-icon icon-name="user" screenreader-text="Screenreader information about the icon goes here, if needed" style="color: #e08915;"></gux-icon>
        <gux-truncate>
          <slot></slot>
        </gux-truncate>
      </mock:shadow-root>
      Michelangelo
    </gux-option-icon>
  </gux-listbox>
</gux-dropdown>
`;
